### ISSUE TITLE
The logprobs in the ChatCompletion's responses is incorrectly reusing the Completions's schema; not following the OpenAI API's spec

### ISSUE THREAD

**Issue by anon998** [vllm-project/vllm#3179]
According to the OpenAI's documentation, the logprobs are returned in a different format when using the Chat Completion API, which is different from the format used in the old Completions one:
https://platform.openai.com/docs/api-reference/chat/create
![image1](https://github.com/vllm-project/vllm/assets/131767832/1a2e48f1-4031-4dcf-8217-885d45c62958)

But vLLM uses the Completion API's way to return the logprobs for the ChatCompletion's responses too.
This is the example chat completion output from the OpenAI API's documentation:
``` json
{
  "id": "chatcmpl-123",
  "object": "chat.completion",
  "created": 1702685778,
  "model": "gpt-3.5-turbo-0125",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "Hello"
      },
      "logprobs": {
        "content": [
          {
            "token": "Hello",
            "logprob": -0.31725305,
            "bytes": [72, 101, 108, 108, 111],
            "top_logprobs": [
              {
                "token": "Hello",
                "logprob": -0.31725305,
                "bytes": [72, 101, 108, 108, 111]
              },
              {
                "token": "Hi",
                "logprob": -1.3190403,
                "bytes": [72, 105]
              }
            ]
          }
        ]
      },
      "finish_reason": "length"
    }
  ],
  "usage": {
    "prompt_tokens": 10,
    "completion_tokens": 1,
    "total_tokens": 11
  },
  "system_fingerprint": null
}
```

And this is what vLLM is returning with commit 901cf4c5:
``` json
{
  "id": "cmpl-feb5333e02ef436e95c09b8f2255e4c0",
  "object": "chat.completion",
  "created": 997761,
  "model": "Qwen/Qwen1.5-72B-Chat-GPTQ-Int4",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "Hello"
      },
      "logprobs": {
        "text_offset": [
          0
        ],
        "token_logprobs": [
          -0.015691734850406647
        ],
        "tokens": [
          "Hello"
        ],
        "top_logprobs": [
          {
            "Hello": -0.015691734850406647,
            "How": -5.140691757202148
          }
        ]
      },
      "finish_reason": "length"
    }
  ],
  "usage": {
    "prompt_tokens": 10,
    "total_tokens": 11,
    "completion_tokens": 1
  }
}
```

Here's how the Completion's LogProbs schema is reused inside the ChatCompletion response:
https://github.com/vllm-project/vllm/blob/901cf4c52bf65472ca13aa4f996d631d00c2228d/vllm/entrypoints/openai/protocol.py#L245-L249

https://github.com/vllm-project/vllm/blob/901cf4c52bf65472ca13aa4f996d631d00c2228d/vllm/entrypoints/openai/protocol.py#L289-L293

And then the `_create_logprobs` function that was made for the Completion API is reused in `serving_chat.py`:
https://github.com/vllm-project/vllm/blob/901cf4c52bf65472ca13aa4f996d631d00c2228d/vllm/entrypoints/openai/serving_chat.py#L241

**Comment by siyuyuan** [comment]
> According to the OpenAI's documentation, the logprobs are returned in a different format when using the Chat Completion API, which is different from the format used in the old Completions one: https://platform.openai.com/docs/api-reference/chat/create ![image1](https://private-user-images.githubusercontent.com/131767832/309770275-1a2e48f1-4031-4dcf-8217-885d45c62958.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk2MjkzODEsIm5iZiI6MTcwOTYyOTA4MSwicGF0aCI6Ii8xMzE3Njc4MzIvMzA5NzcwMjc1LTFhMmU0OGYxLTQwMzEtNGRjZi04MjE3LTg4NWQ0NWM2Mjk1OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwNVQwODU4MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NDYyNjAxMmI2MTNhNDNhMzkxMjU5MTUxZDVhNDJjZDQ3OTYxNzRkMTVmM2VjZDBjN2JjZWI3YWExZTE3YTQzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.cTEfGbIDgBgiP4YwZ1Q2db59zQGRwWQHq3P48Xbjvp0)
> 
> But vLLM uses the Completion API's way to return the logprobs for the ChatCompletion's responses too. This is the example chat completion output from the OpenAI API's documentation:
> 
> ```json
> {
>   "id": "chatcmpl-123",
>   "object": "chat.completion",
>   "created": 1702685778,
>   "model": "gpt-3.5-turbo-0125",
>   "choices": [
>     {
>       "index": 0,
>       "message": {
>         "role": "assistant",
>         "content": "Hello"
>       },
>       "logprobs": {
>         "content": [
>           {
>             "token": "Hello",
>             "logprob": -0.31725305,
>             "bytes": [72, 101, 108, 108, 111],
>             "top_logprobs": [
>               {
>                 "token": "Hello",
>                 "logprob": -0.31725305,
>                 "bytes": [72, 101, 108, 108, 111]
>               },
>               {
>                 "token": "Hi",
>                 "logprob": -1.3190403,
>                 "bytes": [72, 105]
>               }
>             ]
>           }
>         ]
>       },
>       "finish_reason": "length"
>     }
>   ],
>   "usage": {
>     "prompt_tokens": 10,
>     "completion_tokens": 1,
>     "total_tokens": 11
>   },
>   "system_fingerprint": null
> }
> ```
> 
> And this is what vLLM is returning with commit [901cf4c](https://github.com/vllm-project/vllm/commit/901cf4c52bf65472ca13aa4f996d631d00c2228d):
> 
> ```json
> {
>   "id": "cmpl-feb5333e02ef436e95c09b8f2255e4c0",
>   "object": "chat.completion",
>   "created": 997761,
>   "model": "Qwen/Qwen1.5-72B-Chat-GPTQ-Int4",
>   "choices": [
>     {
>       "index": 0,
>       "message": {
>         "role": "assistant",
>         "content": "Hello"
>       },
>       "logprobs": {
>         "text_offset": [
>           0
>         ],
>         "token_logprobs": [
>           -0.015691734850406647
>         ],
>         "tokens": [
>           "Hello"
>         ],
>         "top_logprobs": [
>           {
>             "Hello": -0.015691734850406647,
>             "How": -5.140691757202148
>           }
>         ]
>       },
>       "finish_reason": "length"
>     }
>   ],
>   "usage": {
>     "prompt_tokens": 10,
>     "total_tokens": 11,
>     "completion_tokens": 1
>   }
> }
> ```
> 
> Here's how the Completion's LogProbs schema is reused inside the ChatCompletion response:
> 
> https://github.com/vllm-project/vllm/blob/901cf4c52bf65472ca13aa4f996d631d00c2228d/vllm/entrypoints/openai/protocol.py#L245-L249
> 
> https://github.com/vllm-project/vllm/blob/901cf4c52bf65472ca13aa4f996d631d00c2228d/vllm/entrypoints/openai/protocol.py#L289-L293
> 
> And then the `_create_logprobs` function that was made for the Completion API is reused in `serving_chat.py`:
> 
> https://github.com/vllm-project/vllm/blob/901cf4c52bf65472ca13aa4f996d631d00c2228d/vllm/entrypoints/openai/serving_chat.py#L241

Hello! How do you get logprob using vllm? Can you please release the code public?

**Comment by siyuyuan** [comment]
Thank you for your response! I have tried use Langchain to get logprobs but it is not work:(
For your reference:
```
chat = ChatOpenAI(model=model_name, temperature=0, model_kwargs={"logprobs": True, "top_logprobs": 1})
content = "XXXXXX"
result = chat.generate([[HumanMessage(content=content)]])
```
I will try your method. Thanks again!

**Comment by siyuyuan** [comment]
@anon998 Sorry to bother you.
```
result = openai.ChatCompletion.create(
                temperature=0,
                model=model,
                messages=[{"role": "user", "content": "Hi!"}],
                max_tokens=500,
                top_logprobs=2,
                logprobs=True,)
```
This is my code. However, the result is:

```
<OpenAIObject chat.completion id=cmpl-d954fdf1cb924488a7fd68eb1f2d7641 at 0x7fe2713f2860> JSON: {
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "message": {
        "content": " Hello! How can I help you today?",
        "role": "assistant"
      }
    }
  ],
  "created": 7253,
  "id": "cmpl-d954fdf1cb924488a7fd68eb1f2d7641",
  "model": "/home/huggingface_models/models--openchat--openchat-3.5-1210/snapshots/e5df841b685e5b5ca11ce142f29c6c731bf087a0",
  "object": "chat.completion",
  "usage": {
    "completion_tokens": 10,
    "prompt_tokens": 19,
    "total_tokens": 29
  }
}

```
Do you know why it is?

**Comment by siyuyuan** [comment]
@anon998 Thank you so much! It works!!

**Comment by SachitS** [comment]
@anon998 have you found a workaround to fix this issue yet, or just patch it on client?

**Comment by DarkLight1337** [comment]
Fixed by #5029
